wallet/manager.go

```diff
@@ -237,32 +270,244 @@ func (m *Manager) UnconfirmedEvents() ([]Event, error) {
+
+		for _, sce := range ephemeral {
+			if inPool[sce.ID] || m.utxosLocked(types.Hash256(sce.ID)) != nil {
+				continue
+			}
+
+			selected = append(selected, sce)
+			inputSum = inputSum.Add(sce.SiacoinOutput.Value)
+			if inputSum.Cmp(amount) >= 0 {
+				break
+			}
+		}
+	}
+
+	if inputSum.Cmp(amount) < 0 {
+		return nil, types.ChainIndex{}, types.ZeroCurrency, ErrInsufficientFunds
+	}
+	m.lockUTXOs(utxoIDs...)
+	return selected, tip, inputSum.Sub(amount), nil
+}
+
+// SelectSiafundElements selects siafund elements from the wallet that sum to
+// at least the given amount. Returns the elements, the element basis, and the
+// change amount.
+func (m *Manager) SelectSiafundElements(walletID ID, amount uint64) ([]types.SiafundElement, types.ChainIndex, uint64, error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	tip := m.chain.Tip()
+	if amount == 0 {
+		return nil, tip, 0, nil
+	}
+
+	knownAddresses := make(map[types.Address]bool)
+	relevantAddr := func(addr types.Address) (bool, error) {
+		if exists, ok := knownAddresses[addr]; ok {
+			return exists, nil
+		}
+		_, err := m.store.WalletAddress(walletID, addr)
+		if errors.Is(err, ErrNotFound) {
+			knownAddresses[addr] = false
+			return false, nil
+		} else if err != nil {
+			return false, err
+		}
+		knownAddresses[addr] = true
+		return true, nil
+	}
+
+	ephemeral := make(map[types.SiafundOutputID]types.SiafundElement)
+	inPool := make(map[types.SiafundOutputID]bool)
+	for _, txn := range m.chain.PoolTransactions() {
+		for _, sfi := range txn.SiafundInputs {
+			inPool[sfi.ParentID] = true
+			delete(ephemeral, sfi.ParentID)
+		}
+		for i, sfo := range txn.SiafundOutputs {
+			exists, err := relevantAddr(sfo.Address)
+			if err != nil {
+				return nil, types.ChainIndex{}, 0, fmt.Errorf("failed to check if address %q is relevant: %w", sfo.Address, err)
+			} else if exists {
+				sfoid := txn.SiafundOutputID(i)
+				ephemeral[sfoid] = types.SiafundElement{
+					ID:            sfoid,
+					StateElement:  types.StateElement{LeafIndex: types.UnassignedLeafIndex},
+					SiafundOutput: sfo,
+				}
+			}
+		}
+	}
+	for _, txn := range m.chain.V2PoolTransactions() {
+		for _, sfi := range txn.SiafundInputs {
+			inPool[sfi.Parent.ID] = true
+			delete(ephemeral, sfi.Parent.ID)
+		}
+		for i, sfo := range txn.SiafundOutputs {
+			exists, err := relevantAddr(sfo.Address)
+			if err != nil {
+				return nil, types.ChainIndex{}, 0, fmt.Errorf("failed to check if address %q is relevant: %w", sfo.Address, err)
+			} else if exists {
+				sfe := txn.EphemeralSiafundOutput(i)
+				ephemeral[sfe.ID] = sfe
+			}
+		}
+	}
+
+	var inputSum uint64
+	var selected []types.SiafundElement
+	var utxoIDs []types.Hash256
+top:
+	for i := 0; ; i++ {
+		utxos, err := m.store.WalletSiafundOutputs(walletID, i, 100)
+		if err != nil {
+			return nil, types.ChainIndex{}, 0, fmt.Errorf("failed to get siafund elements: %w", err)
+		} else if len(utxos) == 0 {
+			return nil, types.ChainIndex{}, 0, ErrInsufficientFunds
```

Comment:
Wouldn't it be useful to give an indication of how much we're missing here? It's probably quite annoying since it can either be in the tx pool or locked but I think `renterd` at one point had something similar where we returned the amounts `locked`, `inpool` and so on.

Reply:
I think the single address wallet does it. I think this is fine as is though. I have no desire to calculate the amount already locked here.